[MRG+1] split  data using _safe_split in _permutaion_test_scorer to fix error…

equialgo opened the pull request:
… when using Pandas DataFrame/Series

Related to issue #5696

Commit:
fixing E221 multiple spaces before operator flake8 warnings (+6 squashed commits)
Squashed commits:
[94fd9f4] split data using _safe_split in _permutaion_test_scorer
[522053b] adding test case test_permutation_test_score_pandas() to check if permutation_test_score plays nice with pandas dataframe/series
[21b23ce] running test_permutation_test_score_pandas on iris data to prevent warnings.
[15a48bf] adding safe_indexing to _shuffle function
[9ea5c9e] adding test case test_permutation_test_score_pandas() to check if permutation_test_score plays nice with pandas dataframe/series
[3cf5e8f] split  data using _safe_split in _permutaion_test_scorer to fix error when using Pandas DataFrame/Series

rvraghav93 commented:
thanks for fixing this!

Could you also add this fix to the `model_selection` module and also add a NRT to `model_selection/tests/test_validation.py`? :)

rvraghav93 commented:
Also could you edit the

PR title to -
`[MRG] split data using _safe_split in _permutaion_test_scorer to fix error when using Pandas DataFrame/Series`

And the PR description to "Fixes #5696"

amueller commented:
yeah a test would be great, using the mock data frame or an acutal dataframe if pandas is installed.

equialgo commented:
First of all, sorry for the late reply.

I added the fix also to model_selection/_validation.py.

Moreover, I added a `test_permutation_test_score_pandas()` test case to model_selection/tests/test_validation.py (based on the `test_cross_val_score_pandas()` test case)

dankessler commented:
Is there additional work that needs to be done before this can be merged? I'm still encountering the issue that this addresses (#5696), which I'm working around by calling .values on my series. Is it just the merge conflict that's hairy? If so, I'm happy to take a crack.

amueller commented:
Wow sorry this one has been lying around for a bit. It needs a rebase and an entry to ``whatsnew.rst`` - and reviews, though that should be reasonably quick.

amueller commented:
This actually fixes another issue, namely using precomputed kernels / distance matrices with ``permutation_score``. It would be great if there was also a regression test for that.
Otherwise looks good to me. Sorry for the slow turn-around.

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -731,3 +731,22 @@ def test_cross_val_predict_sparse_prediction():
+        check_df = lambda x: isinstance(x, InputFeatureType)
+        check_series = lambda x: isinstance(x, TargetType)
+        clf = CheckingClassifier(check_X=check_df, check_y=check_series)
+        permutation_test_score(clf, X_df, y_ser)
```
Ideally we should check that, for fixed `random_state`, results are identical regardless of input type.

equialgo replied:
At the moment, such input tests are also not present for the other validation methods, there is only: `test_cross_val_score_pandas()`, `test_cross_val_predict_pandas()`.

I wonder is that really a necessity?

jnothman commented:
@equialgo Sorry for the extreme delay. Let us know if you have time soon to complete the changes.

equialgo commented:
Hi guys, I rebased the fix it should be relatively easy to merge now (if it passes the checks).

@amueller  I do not really get why this also fixes the issue using precomputed kernels / distance matrices with `permutation_score`. If it is trivial please elaborate, otherwise it might be better to add an new ticket for that one.

jnothman commented:
You've done something strange with your commit history. Perhaps a rebase is in order.

equialgo commented:
@jnothman oops, kinda messed-up there! Re-did the rebase; commit history should be okay now.

GaelVaroquaux commented:
+1 to merge this guy (there is a flake8 failure, but it's because of lambda expressions, and here they seem legit).

Let's try to merge this fast, it's been lying around for too long.

GaelVaroquaux commented:
@raghavrv : are you +1 for merge? If so, let's merge

raghavrv commented:
I think this needs a whatsnew entry? I can send a quick PR after this gets merged if you want. Otherwise LGTM... That pep8 error needs to be added to ignore list...

raghavrv reviewed (approved):
Pending whatsnew... (Feel free to merge and later add a whatsnew...)

raghavrv commented:
Okay... I'm merging this... Il make a whatsnew entry for this and another PR soon...

raghavrv commented:
Thanks @equialgo!

equialgo commented:
You guys thanks for merging!
